### PR TITLE
Implemented basic inventory handling...

### PR DESCRIPTION
**Pull request by Doridian**
...and currently held item for inventories.

### CODE REVIEW

**Comment by RoyAwesome** [comment]
can you update the pull to these guidelines? http://forums.spout.org/threads/license-and-pull-requests.643/ 

(Notably 1 commit and the signoff message)

File: src/main/java/org/spout/server/SpoutBlock.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	public SpoutBlock(World world, int x, int y, int z, BlockMaterial material) {
+		this(world, x, y, z);
+		setBlockMaterial(material);
+	}
+	
+	@Override
+	public BlockMaterial setBlockMaterial(BlockMaterial material) {
+		setBlockId(material.getId());
```
Comment (Raphfrk) [inline src/main/java/org/spout/server/SpoutBlock.java:27]: This should probably also set the block's data and auxData.

We haven't implemented that yet though.

I am planing to add a BlockFullState class that will have the 3 pieces of info.

File: src/main/java/org/spout/server/SpoutBlock.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	@Override
+	public short setBlockId(short id) {
+		BlockMaterial newMat = MaterialData.getBlock(id);
+		if(newMat != null) {
+			material = newMat;
+		}
+		return material.getId();
+	}
```
Comment (Raphfrk) [inline src/main/java/org/spout/server/SpoutBlock.java]: Could be a NPE here, if material is null?

File: src/main/java/org/spout/server/SpoutChunk.java [hunk 3]
```diff
@@ -342,6 +342,9 @@ public void populate(boolean force) {
 			}
 		}
 		populated.set(true);
+		if(getRegion() instanceof SpoutRegion) {
+			((SpoutRegion) getRegion()).onChunkPopulated(this);
+		}
```
Comment (Raphfrk) [inline src/main/java/org/spout/server/SpoutChunk.java]: I think @narrowtux  already did chunk population?

**Comment by Raphfrk** [comment]
I would also like it updated to the guidelines too.

File: src/main/java/org/spout/server/SpoutRegion.java [hunk 4]
```diff
@@ -78,6 +82,15 @@
 	 */
 	protected final EntityManager entityManager = new EntityManager();
 	
+	/**
+	 * Holds all not populated chunks
+	 */
+	protected Set<Chunk> nonPopulatedChunks = new HashSet<Chunk>();
```
Comment (Raphfrk) [inline src/main/java/org/spout/server/SpoutRegion.java]: This should probably be a ConcurrentHashMap so as to be thread safe

File: src/main/java/org/spout/server/entity/SpoutEntity.java [hunk 5]
```diff
@@ -270,6 +271,26 @@ public void setData(String key, Externalizable value) {
...
+	}
+
+	private Inventory inventory;
+	@Override
+	public Inventory getInventory() {
+		if(inventory == null) {
+			SpoutDatatableObject obj = (SpoutDatatableObject)getData("inventory");
+			if(obj == null) {
+				inventory = new Inventory(getInventorySize());
+				setData("inventory", inventory);
```
Comment (Raphfrk) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: Hmm, so all entities would have an inventory?  Maybe there should be a (boolean create) parameter to indicate if one should be created.

Comment (Doridian) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: Could also just override the getInventorySize() method so it returns 0 (and then implement if into getInventory() that returns null in such cases?)

File: src/main/java/org/spout/server/SpoutBlock.java (hunk 1 above)
Comment (Raphfrk) [inline src/main/java/org/spout/server/SpoutBlock.java:27]: I have added the class to SpoutAPI, so you can get all 3 pieces of info at once.  You will also be able to set all 3 at once.

File: src/main/java/org/spout/server/entity/SpoutEntity.java (hunk 5 above)
Comment (Raphfrk) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: Another thought is that your method causes a string lookup every time.

DataTables allow an id to be used as an alternative.  

Something like:

```
static volatile inventoryId;
static volatile inventoryIdSet = false;
...
...
...
if (!inventoryIdSet) {
    inventoryId = stringMap.register("inventory");
    inventoryIdSet = true;
}
```

There would need to be a way to get the DataTable string map though.  @RoyAwesome may have set up how to do it.

Comment (RoyAwesome) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: Yeah, I'll do that now, it's a good idea.

Comment (Doridian) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: Also my method would only cause a lookup once per entity (when it first stores it into the "local" variable inventory)

Comment (Raphfrk) [inline src/main/java/org/spout/server/entity/SpoutEntity.java:304]: True, I was just aiming to get the principle of using the ids out there :)

**Pull request merged**